include/deal.II/matrix_free/tools.h

```diff
@@ -443,7 +443,189 @@ namespace MatrixFreeTools
                      }),
               locally_relevant_constrains.end());
 
-            // STEP 2c: translate COO to CRS
+            // STEP 2c: apply hanging-node constraints
+            if (dof_info.hanging_node_constraint_masks.size() > 0)
+              {
+                const auto mask =
+                  dof_info
+                    .hanging_node_constraint_masks[(cell * n_lanes + v) *
+                                                     n_fe_components +
+                                                   first_selected_component];
+
+                // cell has hanging nodes
+                if (mask != 0)
+                  {
+                    // check if hanging node internpolation matrix has been set
+                    // up
+                    if (locally_relevant_constrains_hn_map.find(mask) ==
+                        locally_relevant_constrains_hn_map.end())
+                      {
+                        // 1) collect hanging-node constraints for cell assuming
+                        // scalar finite element
+                        AlignedVector<VectorizedArrayType> values_dofs(
+                          dofs_per_component);
+
+                        std::array<unsigned int, VectorizedArrayType::size()>
+                          constraint_mask;
+                        constraint_mask[0] = mask;
+
+                        std::vector<
+                          std::tuple<unsigned int, unsigned int, Number>>
+                          locally_relevant_constrains_hn;
+
+                        for (unsigned int i = 0; i < dofs_per_component; ++i)
+                          {
+                            for (unsigned int j = 0; j < dofs_per_component;
+                                 ++j)
+                              values_dofs[j][0] = static_cast<Number>(i == j);
+
+                            dealii::internal::FEEvaluationHangingNodesFactory<
+                              dim,
+                              Number,
+                              VectorizedArrayType>::apply(1,
+                                                          phi.get_shape_info()
+                                                            .data.front()
+                                                            .fe_degree,
+                                                          phi,
+                                                          false,
+                                                          constraint_mask,
+                                                          values_dofs.data());
+
+                            for (unsigned int j = 0; j < dofs_per_component;
+                                 ++j)
+                              if (1e-10 <
+                                  std::abs(values_dofs[j][0] &&
```

Comment:
I think there is something wrong with the parentheses here.

Reply:
I just noticed the same thing and put up a patch in #12619.

This is very weird. The OSX testers both fail on other branches but they both got green checkmarks here.

Reply:
I only realised because I accidentally built the examples and clang raised a warning (-Wabsolute-value).